flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java

```diff
@@ -364,25 +365,29 @@ public void addOperatorCoordinator(SerializedValue<OperatorCoordinator.Provider>
 	 * @param grp The slot sharing group to associate the vertex with.
 	 */
 	public void setSlotSharingGroup(SlotSharingGroup grp) {
+		checkNotNull(grp);
+
 		if (this.slotSharingGroup != null) {
 			this.slotSharingGroup.removeVertexFromGroup(this.getID(), this.getMinResources());
 		}
 
+		grp.addVertexToGroup(this.getID(), this.getMinResources());
 		this.slotSharingGroup = grp;
-		if (grp != null) {
-			grp.addVertexToGroup(this.getID(), this.getMinResources());
-		}
 	}
 
 	/**
 	 * Gets the slot sharing group that this vertex is associated with. Different vertices in the same
-	 * slot sharing group can run one subtask each in the same slot. If the vertex is not associated with
-	 * a slot sharing group, this method returns {@code null}.
+	 * slot sharing group can run one subtask each in the same slot.
 	 *
-	 * @return The slot sharing group to associate the vertex with, or {@code null}, if not associated with one.
+	 * @return The slot sharing group to associate the vertex with
 	 */
-	@Nullable
 	public SlotSharingGroup getSlotSharingGroup() {
+		if (slotSharingGroup == null) {
+			// create a new slot sharing group for this vertex if it was in no other slot sharing group.
+			// this should only happen in testing cases at the moment because production code path will
+			// always set a value to it before used
```

Comment:
I think so, there are 200+ usages of JobVertex constructors, distributed in tens of tests.

Reply:
ok, it is a pity that `JobVertex` is mutable and we do not have a separate builder for `JobVertex`.
is this `if` basically not the same as just initialising `slotSharingGroup = new SlotSharingGroup()` in the constructor?
this should be also inline with the previous behaviour where all vertexes with null-group get into a separate group

Reply:
It's not the same because we need to invoke `setSlotSharingGroup()` which relies on `getMinResources` of the vertex.
The `getMinResources`, however, is set after the vertex is constructed.

Reply:
Alright, I suggest to consider it as a follow-up cleanup. We can introduce later a non-null SSG check in the getter to see how many tests fail.